lib/rucio/api/replica.py

```diff
@@ -85,36 +86,54 @@ def declare_bad_file_replicas(replicas, reason, issuer, vo='def', force=False, *
     if not replicas:
         return {}
 
-    kwargs = {}
-    rse_map = {}                # RSE name -> RSE id
-    if not permission.has_permission(issuer=issuer, vo=vo, action='declare_bad_file_replicas', kwargs=kwargs, session=session):
-        raise exception.AccessDenied('Account %s can not declare bad replicas' % (issuer))
-
-    issuer = InternalAccount(issuer, vo=vo)
-
     # make sure all elements are either strings or dicts, without mixing
-    type_ = type(replicas[0])
-    if any(not isinstance(r, type_) for r in replicas):
+
+    as_pfns = isinstance(replicas[0], str)
+    if any(isinstance(r, str) != as_pfns for r in replicas):
         raise exception.InvalidType('The replicas must be specified either as a list of PFNs (strings) or list of dicts')
 
+    rse_map = {}            # name -> id
+    rse_id_to_name = {}     # id -> name
+
+    issuer = InternalAccount(issuer, vo=vo)
+
     replicas_lst = replicas
-    if type_ is dict:
+    rse_ids_to_check = set()
+    if as_pfns:
+        scheme, rses_for_replicas, unknowns = get_pfn_to_rse(replicas, vo=vo)
+        if unknowns:
+            raise exception.ReplicaNotFound("Not all replicas found")
+        rse_ids_to_check = set(rses_for_replicas.keys())
+    else:
+        # repliac are given as dicts {name, scope, rse}
         replicas_lst = []           # need to create new list to convert replica["scope"] from strings to InternalScope objects
         for r in replicas:
             if "name" not in r or "scope" not in r or ("rse" not in r and "rse_id" not in r):
                 raise exception.InvalidType('The replica dictionary must include scope and either rse (name) or rse_id')
-            scope = InternalScope(r['scope'], vo=vo)
-            rse_id = r.get("rse_id") or rse_map.get(r['rse'])
-            if rse_id is None:
-                rse = r["rse"]
-                rse_map[rse] = rse_id = get_rse_id(rse=rse, vo=vo, session=session)
+            rse_id = r.get("rse_id")
+            rse = r.get("rse")
+            rse_id = rse_id or rse_map.get(rse) or get_rse_id(rse=rse, vo=vo, session=session)
+            rse = rse or rse_id_to_name.get(rse_id) or get_rse_name(rse_id, session=session)
+            rse_map[rse] = rse_id
+            rse_id_to_name[rse_id] = rse
             replicas_lst.append({
-                "scope": scope,
+                "scope": InternalScope(r['scope'], vo=vo),
                 "rse_id": rse_id,
                 "name": r["name"]
             })
+            rse_ids_to_check.add(rse_id)
 
-    rse_id_to_name = invert_dict(rse_map)   # RSE id -> RSE name
+    for rse_id in rse_ids_to_check:
+        try:
```

Comment:
Why do we need the try/except block here?

Reply:
to handle exceptions caused by potentially buggy implementation of the permission function.

Perhaps it would be better to assume that if the permissions function raises a exception, the answer is "no the client does not have the permission"

Reply:
@bari12 Just to take the opportunity here. In general, we shall modify the interface of the permission module function to be able to raise exceptions or at least pass a message that can be relayed to the client, instead of "`user` does not have permission for `function`". 
As our permission functions become more complex, in certain situations, we would like to return a specific message on why the permission was denied. What do you think?

Reply:
@dynamic-entropy I think that makes sense, but we would have to adapt it coherently, also having the policy modules in mind.
@ivmfnal Ok, it's just a bit "out-of-sync" since we do not do this anywhere else with permissions, I am not against it, but the way it is done it might rather hide things going wrong in the permission module.

Reply:
I think I should revert to the original behavior - not try to catch exceptions raised by the permissions function and let them propagate upwards so that they can be handled as they are - with the type and whatever messages are attached.